### DISCUSSION TITLE
/api route not blocking CORS

### DISCUSSION THREAD

**Discussion by initeshjain** [vercel/next.js#59713] [Help] [answered]
### Summary

I have a website `tiktook.in` which is running on `"next": "13.1.1"`, uses pages structure and /api route.

NextJS should block CORS requests by default but it is not.
I have set up `next.config.js` and `vercel.json` as below. But still, I can hit the API endpoint with the browser and Postman as well. I want to disable CORS and avoid allowing it from Postman and browsers directly.

Example: As per Cors policy, this link: `https://www.tiktook.in/api/getItems?page=blogs`, should not be reachable directly.

```
// next.config.js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "*",
        port: "",
      },
    ],
  },
  async headers() {
    console.log("Executing headers function");
    return [
      {
        // matching all API routes
        source: "/api/(.*)",
        headers: [
          { key: "Access-Control-Allow-Credentials", value: "true" },
          { key: "Access-Control-Allow-Origin", value: "http://tiktook.in,https://tiktook.in,http://www.tiktook.in,https://www.tiktook.in,tiktook.in,www.tiktook.in" },
          { key: "Access-Control-Allow-Methods", value: "GET,DELETE,PATCH,POST,PUT" },
          { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" },
        ]
      }
    ]
  }
};

module.exports = nextConfig;
```

```
// vercel.json
{
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        {
          "key": "Access-Control-Allow-Credentials",
          "value": "true"
        },
        {
          "key": "Access-Control-Allow-Origin",
          "value": "http://tiktook.in,https://tiktook.in,http://www.tiktook.in,https://www.tiktook.in,tiktook.in,www.tiktook.in"
        },
        {
          "key": "Access-Control-Allow-Methods",
          "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT"
        },
        {
          "key": "Access-Control-Allow-Headers",
          "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version"
        }
      ]
    }
  ]
}
```

**Comment by initeshjain** [comment]
Any help will be appreciated.

**Comment by icyJoseph** [comment]
Did you fix this? If I try to get that endpoint from say, the next.js docs website, it fails on CORS, as you expect it to do.

<img width="1364" alt="Screenshot 2024-01-07 at 12 37 28" src="https://github.com/vercel/next.js/assets/21013447/02c5b033-021c-4109-9f9b-f12bbba4e8ba">

I think you might be a bit confused, as to why Postman seems to allow it? 

Maybe these two answers can clarify that:

- https://stackoverflow.com/a/45764496
- https://stackoverflow.com/a/36486188

**Reply by initeshjain** [reply]
But, if I just access the URL: `https://www.tiktook.in/api/getItems?page=blogs` from a browser as a get-call, it returns me the data. Is this accepted behavior? I thought, direct calls would also get blocked. Like:

![image](https://github.com/vercel/next.js/assets/47496910/b581115e-edec-4ebe-88e2-5a85bec8c7a3)

**Reply by icyJoseph** [reply, marked as answer]
Not really, right? CORS is to tell a browser, that, my resources in `example.com`, should only be allowed in certain domains, for example, `mydomain.com`, so that if `evil.com` tries to request stuff from `example.com`, it'll be blocked by CORS policy.

- https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

You might want to check if there's an origin header in the request, a referrer, or some other kind of Cookie based approach, to forbid access to that endpoint.

**Reply by initeshjain** [reply]
Oh, I am getting a referrer, I can use that to block direct calls as well.
Thanks for the explaination.

**Reply by icyJoseph** [reply]
No worries, be mindful that, these things can be spoofed, so it all depends on how much you care about protecting this route.

**Reply by initeshjain** [reply]
yup, got it.
